ext/mbstring/mbstring.c

```diff
@@ -4800,11 +4809,355 @@ finish_up_remaining_bytes: ;
+	 * and Daniel Lemire.
+	 * Ref: https://arxiv.org/pdf/2010.03090.pdf
+	 *
+	 * Most types of invalid UTF-8 text can be detected by examining pairs of
+	 * successive bytes. Specifically:
+	 *
+	 * • Overlong 2-byte code units start with 0xC0 or 0xC1.
+	 *   No valid UTF-8 string ever uses these byte values.
+	 * • Overlong 3-byte code units start with 0xE0, followed by a byte < 0xA0.
+	 * • Overlong 4-byte code units start with 0xF0, followed by a byte < 0x90.
+	 * • 5-byte or 6-byte code units, which should never be used, start with
+	 *   0xF8-FE.
+	 * • A codepoint value higher than U+10FFFF, which is the highest value for
+	 *   any Unicode codepoint, would either start with 0xF4, followed by a
+	 *   byte >= 0x90, or else would start with 0xF5-F7, followed by any value.
+	 * • A codepoint value from U+D800-DFFF, which are reserved and should never
+	 *   be used, would start with 0xED, followed by a byte >= 0xA0.
+	 * • The byte value 0xFF is also illegal and is never used in valid UTF-8.
+	 *
+	 * To detect all these problems, for each pair of successive bytes, we do
+	 * table lookups using the high nibble of the first byte, the low nibble of
+	 * the first byte, and the high nibble of the second byte. Each table lookup
+	 * retrieves a bitmask, in which each 1 bit indicates a possible invalid
+	 * combination; AND those three bitmasks together, and any 1 bit in the result
+	 * will indicate an actual invalid byte combination was found.
+	 */
+
+#define BAD_BYTE 0x1
+#define OVERLONG_2BYTE 0x2
+#define _1BYTE (BAD_BYTE | OVERLONG_2BYTE)
+#define OVERLONG_3BYTE 0x4
+#define SURROGATE 0x8
+#define OVERLONG_4BYTE 0x10
+#define INVALID_CP 0x20
+
+	/* Each of these are 16-entry tables, repeated twice; this is required by the
+	 * VPSHUFB instruction which we use to perform 32 table lookups in parallel
+	 * The first entry is for 0xF, the second is for 0xE, and so on down to 0x0
+	 *
+	 * So, for example, notice that the 4th entry in the 1st table is OVERLONG_2BYTE;
+	 * that means that high nibble 0xC is consistent with the byte pair being part of
+	 * an overlong 2-byte code unit */
+	const __m256i bad_hi_nibble2 = _mm256_set_epi8(
+		BAD_BYTE | OVERLONG_4BYTE | INVALID_CP, OVERLONG_3BYTE | SURROGATE, 0, OVERLONG_2BYTE,
+		0, 0, 0, 0,
+		0, 0, 0, 0,
+		0, 0, 0, 0,
+		BAD_BYTE | OVERLONG_4BYTE | INVALID_CP, OVERLONG_3BYTE | SURROGATE, 0, OVERLONG_2BYTE,
+		0, 0, 0, 0,
+		0, 0, 0, 0,
+		0, 0, 0, 0);
+	const __m256i bad_lo_nibble2 = _mm256_set_epi8(
+		BAD_BYTE, BAD_BYTE, BAD_BYTE | SURROGATE, BAD_BYTE,
+		BAD_BYTE, BAD_BYTE, BAD_BYTE, BAD_BYTE,
+		BAD_BYTE, BAD_BYTE, BAD_BYTE, INVALID_CP,
+		0, 0, OVERLONG_2BYTE, OVERLONG_2BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE,
+		BAD_BYTE, BAD_BYTE, BAD_BYTE | SURROGATE, BAD_BYTE,
+		BAD_BYTE, BAD_BYTE, BAD_BYTE, BAD_BYTE,
+		BAD_BYTE, BAD_BYTE, BAD_BYTE, INVALID_CP,
+		0, 0, OVERLONG_2BYTE, OVERLONG_2BYTE | OVERLONG_3BYTE | OVERLONG_4BYTE);
+	const __m256i bad_hi_nibble = _mm256_set_epi8(
```

Comment:
I'm confused what's the purpose of this table again?

Reply:
The three 32-byte tables are used for looking up the "possible bad 2-byte combination" bitmasks for the high nibble of the first byte, low nibble of the first byte, and high nibble of the second byte. If any bit is set in all 3 of the resulting bitmasks, then it means all three of those nibbles match them same bad 2-byte combination... meaning the string is invalid.

Each of the 32-byte tables are actually the same 16-byte table, repeated twice. This is because the AVX2 instruction which is used for table lookups splits a 256-bit YMM register into two halves, and uses the first 16 bytes as the lookup table for the first half, and the second 16 bytes as the lookup table for the second half.

We don't really want this behavior, as for our use case, we want to use the *same* lookup table for all the bytes in the entire register, but it's how the instruction works.

Reply:
Right okay, thanks for the explanation. :)